kdc/process.c

```diff
@@ -47,89 +200,85 @@ krb5_kdc_update_time(struct timeval *tv)
 	_kdc_now = *tv;
 }
 
+
+#define EXTEND_REQUEST_T(LHS, RHS) do {			\
+	RHS = realloc(LHS, sizeof(*RHS));		\
+	if (!RHS)					\
+	    return krb5_enomem((LHS)->context);		\
+	LHS = (void *)RHS;				\
+	memset(((char *)LHS) + sizeof(*LHS),		\
+	       0x0,					\
+	       sizeof(*RHS) - sizeof(*LHS));		\
+    } while (0)
+
 static krb5_error_code
-kdc_as_req(krb5_context context,
-	   krb5_kdc_configuration *config,
-	   krb5_data *req_buffer,
-	   krb5_data *reply,
-	   const char *from,
-	   struct sockaddr *addr,
-	   int datagram_reply,
-	   int *claim)
+kdc_as_req(kdc_request_t *rptr, int *claim)
 {
-    struct kdc_request_desc r;
+    astgs_request_t r;
     krb5_error_code ret;
     size_t len;
 
-    memset(&r, 0, sizeof(r));
+    EXTEND_REQUEST_T(*rptr, r);
 
-    ret = decode_AS_REQ(req_buffer->data, req_buffer->length, &r.req, &len);
+    ret = decode_AS_REQ(r->request.data, r->request.length, &r->req, &len);
     if (ret)
 	return ret;
 
-    r.context = context;
-    r.config = config;
-    r.request.data = req_buffer->data;
-    r.request.length = req_buffer->length;
-
+    r->reqtype = "AS-REQ";
+    r->use_request_t = 1;
     *claim = 1;
 
-    ret = _kdc_as_rep(&r, reply, from, addr, datagram_reply);
-    free_AS_REQ(&r.req);
+    ret = _kdc_as_rep(r);
+    free_AS_REQ(&r->req);
     return ret;
 }
 
 
 static krb5_error_code
-kdc_tgs_req(krb5_context context,
-	    krb5_kdc_configuration *config,
-	    krb5_data *req_buffer,
-	    krb5_data *reply,
-	    const char *from,
-	    struct sockaddr *addr,
-	    int datagram_reply,
-	    int *claim)
+kdc_tgs_req(kdc_request_t *rptr, int *claim)
 {
+    astgs_request_t r;
     krb5_error_code ret;
-    KDC_REQ req;
     size_t len;
 
-    ret = decode_TGS_REQ(req_buffer->data, req_buffer->length, &req, &len);
+    EXTEND_REQUEST_T(*rptr, r);
+
+    ret = decode_TGS_REQ(r->request.data, r->request.length, &r->req, &len);
     if (ret)
 	return ret;
 
+    r->reqtype = "TGS-REQ";
+    r->use_request_t = 1;
     *claim = 1;
 
-    ret = _kdc_tgs_rep(context, config, &req, reply,
-		       from, addr, datagram_reply);
-    free_TGS_REQ(&req);
+    ret = _kdc_tgs_rep(r);
+    free_TGS_REQ(&r->req);
     return ret;
 }
 
 #ifdef DIGEST
 
 static krb5_error_code
-kdc_digest(krb5_context context,
-	   krb5_kdc_configuration *config,
-	   krb5_data *req_buffer,
-	   krb5_data *reply,
-	   const char *from,
-	   struct sockaddr *addr,
-	   int datagram_reply,
-	   int *claim)
+kdc_digest(kdc_request_t *rptr, int *claim)
 {
+    kdc_request_t r;
     DigestREQ digestreq;
     krb5_error_code ret;
     size_t len;
 
-    ret = decode_DigestREQ(req_buffer->data, req_buffer->length,
+    /* XXXrcd: we haven't implemented the other stuff here */
```

Comment:
I've put this in as 29755c0, a separate commit in the same PR.

Reply:
Oh, above comment is incorrect. That was for something else.

Reply:
No, so that's keeping track of my state of mind and I should likely remove the comment. I provided a mechanism for services to "sign up" for the kdc_request_t, it is:
            r->use_request_t = 0;
As you can see, if that's zero, then process.c will not generate a log line for the service under the assumption that it has not correctly populated the requisite data.